File: java-kata/src/main/java/org/kata/sbe/loan/domain/ApplicationResult.java
package org.kata.sbe.loan.domain;

import java.util.Collections;
import java.util.List;

public class ApplicationResult {
    private final ApplicationStatus status;
    private final List<DeclineReason> declineReasons;

    private ApplicationResult(ApplicationStatus status, List<DeclineReason> declineReasons) {
        this.status = status;
        this.declineReasons = Collections.unmodifiableList(declineReasons);
    }

    public static ApplicationResult approved() {
        return new ApplicationResult(ApplicationStatus.APPROVED, Collections.emptyList());
    }

    public static ApplicationResult declined(List<DeclineReason> reasons) {
        return new ApplicationResult(ApplicationStatus.DECLINED, reasons);
    }

    // Getters
    public ApplicationStatus getStatus() { return status; }
    public List<DeclineReason> getDeclineReasons() { return declineReasons; }
    public boolean isApproved() { return status == ApplicationStatus.APPROVED; }
}

File: dotnet-kata/Domain/ApplicationResult.cs
namespace Domain;

public class ApplicationResult
{
    public ApplicationStatus Status { get; }
    public IReadOnlyList<DeclineReason> DeclineReasons { get; }
    public bool IsApproved => Status == ApplicationStatus.Approved;

    private ApplicationResult(ApplicationStatus status, IReadOnlyList<DeclineReason> declineReasons)
    {
        Status = status;
        DeclineReasons = declineReasons;
    }

    public static ApplicationResult Approved() => 
        new(ApplicationStatus.Approved, Array.Empty<DeclineReason>());

    public static ApplicationResult Declined(IEnumerable<DeclineReason> reasons) => 
        new(ApplicationStatus.Declined, reasons.ToList().AsReadOnly());
}